KAFKA-5127 Replace pattern matching with foreach where the case None …

baluchicken opened the pull request:
@ijuma plz review. This one is not complete because KafkaController, AdminUtils and ReplicaStateMachine has some of these but for those I will wait until the KAFKA-5028 and KAFKA-5103 are merged.

baluchicken committed:
KAFKA-5127 Replace pattern matching with foreach where the case None is unused

vahidhashemian commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -705,9 +701,9 @@ class ZkUtils(val zkClient: ZkClient,
-        case None =>
+      ReplicationUtils.getLeaderIsrAndEpochForPartition(this, topicAndPartition.topic,
+        topicAndPartition.partition).foreach { leaderIsrAndControllerEpoch =>
+        ret.put(topicAndPartition, leaderIsrAndControllerEpoch)
```
I think this line needs to be indented. I actually prefer the old block format since it's more readable (even if the first line is long).

baluchicken replied:
Thanks, I agree will create a new commit to fix this.

vahidhashemian reviewed (changes requested):
Thanks for refactoring the code. While waiting the jenkins tests results I did have a single comment. Otherwise, LGTM.

baluchicken committed:
Change line indentation on zkUtils

vahidhashemian commented on core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala:
```diff
@@ -218,9 +218,8 @@ private[kafka] class ZookeeperConsumerConnector(val config: ConsumerConfig,
-            case None =>
+          fetcher.foreach { f =>
+            f.stopConnections
           }
```
Why not make this even more compact `fetcher.foreach(_.stopConnections)`?

vahidhashemian commented on core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala:
```diff
@@ -831,10 +828,8 @@ private[kafka] class ZookeeperConsumerConnector(val config: ConsumerConfig,
+      fetcher.foreach { f =>
           f.startConnections(allPartitionInfos, cluster)
-        case None =>
       }
```
Also this: `fetcher.foreach(_.startConnections(allPartitionInfos, cluster))`

vahidhashemian commented on core/src/main/scala/kafka/metrics/KafkaMetricsGroup.scala:
```diff
@@ -68,10 +68,8 @@ trait KafkaMetricsGroup extends Logging {
+    tagsName.foreach { tn =>
         nameBuilder.append(",").append(tn)
-      case None =>
     }
```
And this: `tagsName.foreach(nameBuilder.append(",").append(_))`

vahidhashemian commented on core/src/main/scala/kafka/network/RequestOrResponseSend.scala:
```diff
@@ -27,10 +27,8 @@ import org.apache.kafka.common.network.NetworkSend
+    request.requestId.foreach { requestId =>
         buffer.putShort(requestId)
-      case None =>
     }
```
And this: `request.requestId.foreach(buffer.putShort(_))`

vahidhashemian commented on core/src/main/scala/kafka/producer/async/DefaultEventHandler.scala:
```diff
@@ -114,9 +114,8 @@ class DefaultEventHandler[K,V](config: ProducerConfig,
-                  case None => // nothing
+                messagesPerBrokerMap.get(topicPartition).foreach { data =>
+                  failedProduceRequests.appendAll(data)
                 }
```
`messagesPerBrokerMap.get(topicPartition).foreach(failedProduceRequests.appendAll(_))`

ijuma replied:
Even:

```scala
messagesPerBrokerMap.get(topicPartition).foreach(failedProduceRequests.appendAll)
```

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsumerOffsetChecker.scala:
```diff
@@ -197,9 +191,8 @@ object ConsumerOffsetChecker extends Logging {
-          case None => // ignore
+        consumerOpt.foreach { consumer =>
+          consumer.close()
         }
```
Here too.

ijuma replied:
`consumerMap.values.flatten.foreach(_.close())`

vahidhashemian commented on core/src/main/scala/kafka/tools/ConsumerOffsetChecker.scala:
```diff
@@ -208,9 +201,8 @@ object ConsumerOffsetChecker extends Logging {
-          case None => // ignore
+        consumerOpt.foreach { consumer =>
+          consumer.close()
         }
```
Here too.

vahidhashemian commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -705,9 +701,8 @@ class ZkUtils(val zkClient: ZkClient,
-        case None =>
+      ReplicationUtils.getLeaderIsrAndEpochForPartition(this, topicAndPartition.topic, topicAndPartition.partition).foreach { leaderIsrAndControllerEpoch =>
+        ret.put(topicAndPartition, leaderIsrAndControllerEpoch)
       }
```
And here.

vahidhashemian commented on core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala:
```diff
@@ -375,13 +375,11 @@ class ConsumerBounceTest extends IntegrationTestHarness with Logging {
-        case None =>
+      maxCloseTimeMs.foreach { ms =>
+        assertTrue("Close took too long " + timeTakenMs, timeTakenMs < ms + closeGraceTimeMs)
       }
```
Here.

vahidhashemian commented on core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala:
```diff
@@ -375,13 +375,11 @@ class ConsumerBounceTest extends IntegrationTestHarness with Logging {
-        case None =>
+      minCloseTimeMs.foreach { ms =>
+        assertTrue("Close finished too quickly " + timeTakenMs, timeTakenMs >= ms)
       }
```
Here.

vahidhashemian commented on core/src/test/scala/unit/kafka/integration/MetricsDuringTopicCreationDeletionTest.scala:
```diff
@@ -59,9 +59,8 @@ class MetricsDuringTopicCreationDeletionTest extends KafkaServerTestHarness with
-           case None =>
+        .headOption.foreach { e =>
+          Metrics.defaultRegistry.removeMetric(e._1)
         }
```
Here.

baluchicken replied:
Sorry, but can you help me with this one? I am not sure how to do this:)

vahidhashemian replied:
Sorry my bad, you can ignore this one. The same compaction cannot be applied here.

vahidhashemian commented on core/src/test/scala/unit/kafka/server/epoch/util/ReplicaFetcherMockBlockingSend.scala:
```diff
@@ -48,9 +48,8 @@ class ReplicaFetcherMockBlockingSend(offsets: java.util.Map[TopicPartition, Epoc
-          case None => //nothing
+        callback.foreach { f =>
+          f()
         }
```
And here.

vahidhashemian reviewed (commented):
Sorry for not catching these in the earlier review. There are some `foreach` blocks that can be further compressed. I mentioned them inline. I hope they make sense.

baluchicken committed:
Refactor KafkaController,ReplicationStateMachine class to use foreach instead of pattern matching, further compressed foreach where it makes sence

baluchicken commented:
Updated the PR, refactored the KafkaController and the ReplicationStateMachine. In my first comment I mentioned the AdminUtils also needs refactoring but it doesn't. I also further compressed the foreach blocks but I left out the ones where we have variable with actual names which I think helps a lot while reading the code. For example (ms, leaderIsrAndControllerEpoch, data).

vahidhashemian commented:
Sounds good to me. Maybe @ijuma or @hachikuji can take a look and take it from here.

BTW, A PR for the transient failure that failed one of the tests was just merged. So if you're going to make changes to the PR please rebase and it should go away.

baluchicken commented:
@vahidhashemian Thanks for the tip, and also thank you for the review:)

baluchicken commented:
@ijuma if you have time can you please take a look? Thanks.

baluchicken commented:
Resolved the conflicts.

baluchicken commented:
Resolved the conflicts.

baluchicken commented:
Retest this plz

baluchicken commented:
Recheck

baluchicken commented:
retest this please

baluchicken commented:
@ijuma if you have time can you please take a look? Thanks.

ijuma commented on core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala:
```diff
@@ -778,9 +775,8 @@ private[kafka] class ZookeeperConsumerConnector(val config: ConsumerConfig,
-        case Some(f) =>
-          f.stopConnections
+      fetcher.foreach { f =>
+          f.stopConnections()
```
Generally, you have to fix the indent in every case where you replace a pattern match with `foreach`.

baluchicken committed:
Further compress foreach blocks

baluchicken committed:
Fix indentation

baluchicken commented:
I fixed the indentation and also compressed some foreach blocks further. Thanks for the review.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5399/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5384/
Test PASSed (JDK 8 and Scala 2.12).

baluchicken commented:
@ijuma if you have time can you please take a look? :)

hachikuji commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -190,15 +189,14 @@ class ReplicaStateMachine(controller: KafkaController) extends Logging {
-                    .format(controllerId, controller.epoch, replicaId, topicAndPartition, currState, targetState))
-                case None => // that means the partition was never in OnlinePartition state, this means the broker never
-                  // started a log for that partition and does not have a high watermark value for this partition
+              // None means the partition was never in OnlinePartition state, this means the broker never
```
This probably comes down to personal preference, but in cases when we need to communicate a non-trivial explanation for the `None` case, I found the pattern matching approach easier to follow because the comment appears in the context of the case it applies to.

baluchicken replied:
@hachikuji maybe you are right:) I am going to refactor those parts of the code. Thanks for the review.

baluchicken committed:
Use pattern matching where non trivial comments are present inside the None case

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6077/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6062/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji reviewed (approved):
Thanks for the patch. LGTM
